File: backend/feed-service/src/main/java/ru/maxigram/backend/feedservice/service/CommentService.java
package ru.maxigram.backend.feedservice.service;

import java.util.Collection;
import java.util.Optional;
import ru.maxigram.backend.feedservice.entity.Comment;
import ru.maxigram.backend.feedservice.entity.Post;

public interface CommentService {

  Comment createComment(String authorId, String text, Post post);

  Optional<Comment> findCommentById(Long id);

  Collection<Comment> findCommentsByPost(Post post);

  Collection<Long> getPostIdsCommentedByAuthor(String authorId);
}

File: src/main/kotlin/org/maxizenit/maxigram/feed/CommentService.kt
package org.maxizenit.maxigram.feed

import org.springframework.stereotype.Service
import java.time.Clock
import java.time.Instant
import java.util.UUID

@Service
class CommentService(
    private val comments: CommentRepository,
    private val posts: PostRepository,
    private val clock: Clock = Clock.systemUTC(),
) {

    fun add(postId: Long, authorId: UUID, text: String): Comment {
        if (!posts.existsById(postId)) throw PostNotFoundException(postId)
        val trimmed = text.trim()
        if (trimmed.isEmpty()) throw InvalidContentException("Comment text must not be blank")
        return comments.insert(postId, authorId, trimmed, Instant.now(clock))
    }

    fun viewsForPost(postId: Long, requesterId: UUID): List<CommentView> {
        if (!posts.existsById(postId)) throw PostNotFoundException(postId)
        return comments.viewsForPost(postId, requesterId)
    }
}
